src/arch/util/machine-smp.C

```diff
@@ -514,48 +512,103 @@ static void CmiDestroyLocks(void)
 
 #if !CMK_SHARED_VARS_UNAVAILABLE
 
-#if __cpp_lib_barrier
-
-using Barrier = std::barrier;
-
-#else
-
-class Barrier {
+class Barrier
+{
 public:
   Barrier(const Barrier&) = delete;
   Barrier& operator=(const Barrier&) = delete;
 
-  explicit Barrier(unsigned int count) :
-    curCount(count), barrierCount(count), curSense(true) {
+  explicit Barrier(unsigned int count)
+      : curCount(count), barrierCount(count), curSense(true)
+  {
   }
 
-  void arrive_and_wait() {
-    std::unique_lock<std::mutex> lock(barrierMutex);
+  void arrive_and_wait()
+  {
     const bool sense = curSense;
 
-    curCount--;
-
-    if (curCount == 0) {
-      curSense = !curSense;
+    // If we're last, reset the count and flip the sense
+    if (--curCount == 0)
+    {
       curCount = barrierCount;
-      barrierCond.notify_all();
+      curSense = !curSense;
       return;
     }
 
-    while (sense == curSense) {
-      barrierCond.wait(lock);
+    // Otherwise we're not the last, so start the progressive spin sequence:
+    // Define how many iterations to spend in each phase. These are arbitrary constants,
+    // they can be tuned for performance, but this should be generally alright.
+    constexpr auto spinIters = 5, pauseIters = 10, yieldIters = 1000;
+
+    // Straight up spin at first.
+    for (int i = 0; i < spinIters; i++)
+    {
+      if (sense != curSense.load(std::memory_order_acquire))
+        return;
+    }
+
+    // If that's taking too long, then start pausing during each spin.
+    for (int i = 0; i < pauseIters; i++)
+    {
+      if (sense != curSense)
+        return;
+      pause();
+    }
+
+    // If we're still in the barrier, then it's time to lay off a bit
+    // and start yielding this thread at the OS level. Without this,
+    // oversubscribed scenarios become extremely costly.
+    while (true)
+    {
+      for (int i = 0; i < yieldIters; i++)
+      {
+        if (sense != curSense)
+          return;
+        repeat_pause<10>();
+      }
+      std::this_thread::yield();
     }
   }
 
 private:
-  std::mutex barrierMutex;
-  std::condition_variable barrierCond;
-  bool curSense;
-  unsigned int curCount;
+  alignas(CMI_CACHE_LINE_SIZE) std::atomic<bool> curSense;
+  alignas(CMI_CACHE_LINE_SIZE) std::atomic<unsigned int> curCount;
   const unsigned int barrierCount;
+
+  void pause() const
```

Comment:
Could be marked `static` since it does not use `this`. May also be a good case for CMI_FORCE_INLINE since the function body is either one or zero instructions.

Reply:
Done.